app/src/main/java/ai/elimu/appstore/synchronization/AppSynchronizationActivity.java

```diff
@@ -138,13 +162,32 @@ protected Void doInBackground(Void... voids) {
                                 application.setLiteracySkills(applicationGson.getLiteracySkills());
                                 application.setNumeracySkills(applicationGson.getNumeracySkills());
                                 application.setApplicationStatus(applicationGson.getApplicationStatus());
-                                if (applicationGson.getApplicationStatus() == ApplicationStatus.ACTIVE) {
-                                    ApplicationVersionGson applicationVersionGson = applicationGson.getApplicationVersions().get(0);
-                                    application.setVersionCode(applicationVersionGson.getVersionCode());
-                                    application.setStartCommand(applicationVersionGson.getStartCommand());
-                                }
                                 applicationDao.update(application);
                                 Timber.i("Updated Application in database with id " + application.getId());
+
+                                if (application.getApplicationStatus() == ApplicationStatus.ACTIVE) {
+                                    // Update ApplicationVersions
+                                    List<ApplicationVersionGson> applicationVersionGsons = applicationGson.getApplicationVersions();
+                                    Timber.i("applicationVersionGsons.size(): " + applicationVersionGsons.size());
+                                    for (ApplicationVersionGson applicationVersionGson : applicationVersionGsons) {
+                                        ApplicationVersion applicationVersion = applicationVersionDao.load(applicationVersionGson.getId());
+
+                                        if (applicationVersion == null) {
+                                            // Update existing ApplicationVersion in database
+                                            applicationVersion = new ApplicationVersion();
+                                            applicationVersion.setId(applicationVersionGson.getId());
+                                            applicationVersion.setApplication(application);
+                                            applicationVersion.setFileSizeInKb(applicationVersionGson.getFileSizeInKb());
+                                            applicationVersion.setFileUrl(applicationVersionGson.getFileUrl());
+                                            applicationVersion.setContentType(applicationVersionGson.getContentType());
+                                            applicationVersion.setVersionCode(applicationVersionGson.getVersionCode());
+                                            applicationVersion.setStartCommand(applicationVersionGson.getStartCommand());
+                                            applicationVersion.setTimeUploaded(applicationVersionGson.getTimeUploaded());
+                                            applicationVersionDao.update(applicationVersion);
```

Comment:
You could reuse the above code for the insert for both insert/update to reduce the number of lines